src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java

```diff
@@ -0,0 +1,229 @@
+  private ByteBuffer buffer = ByteBuffer.allocate(ONESHOT_THRESHOLD);
+
+  private long hasher = -1;
+  private boolean isDone;
+
+  public Blake3MessageDigest() {
+    super("BLAKE3");
+    isDone = false;
+  }
+
+  private void flush() {
+    if (hasher == -1) {
+      hasher = Blake3JNI.allocate_and_initialize_hasher();
+    }
+
+    if (buffer.position() > 0) {
+      Blake3JNI.blake3_hasher_update(hasher, buffer.array(), buffer.position());
+      buffer.clear();
+    }
+  }
+
+  public void engineUpdate(byte[] data, int offset, int length) {
+    while (length > 0) {
+      int numToCopy = Math.min(length, buffer.remaining());
+      buffer.put(data, offset, numToCopy);
+      length -= numToCopy;
+      offset += numToCopy;
+
+      if (buffer.remaining() == 0) {
+        flush();
+      }
+    }
+  }
+
+  public void engineUpdate(byte[] data) {
+    engineUpdate(data, 0, data.length);
+  }
+
+  public void engineUpdate(byte b) {
+    engineUpdate(new byte[] {b});
+  }
+
+  private byte[] getOutput(int outputLength) throws IllegalArgumentException {
+    byte[] retByteArray = new byte[outputLength];
+
+    checkState(!isDone);
+    isDone = true;
+
+    if (hasher == -1) {
+      // If no flush has happened yet; oneshot this.
+      Blake3JNI.oneshot(buffer.array(), buffer.position(), retByteArray, outputLength);
+      buffer.clear();
+    } else {
+      flush();
+      Blake3JNI.blake3_hasher_finalize_and_close(hasher, retByteArray, outputLength);
+      hasher = -1;
+    }
+    return retByteArray;
+  }
+
+  public Object clone() throws CloneNotSupportedException {
+    throw new CloneNotSupportedException();
+  }
+
+  public void engineReset() {
+    if (hasher != -1) {
+      Blake3JNI.blake3_hasher_close(hasher);
+      hasher = -1;
+    }
+    buffer.clear();
+    isDone = false;
+  }
+
+  public void engineUpdate(ByteBuffer input) {
+    if (input.hasArray()) {
+      engineUpdate(input.array());
+    } else {
+      byte[] bufCopy = new byte[input.position()];
+      input.get(bufCopy);
+      engineUpdate(bufCopy);
+    }
+  }
+
+  public int engineGetDigestLength() {
+    return OUT_LEN;
+  }
+
+  public byte[] engineDigest() {
+    byte[] digestBytes = getOutput(OUT_LEN);
+    return digestBytes;
+  }
+
+  public int engineDigest(byte[] buf, int off, int len) throws DigestException {
+    if (len < OUT_LEN) {
+      throw new DigestException("partial digests not returned");
+    }
+    if (buf.length - off < OUT_LEN) {
+      throw new DigestException("insufficient space in the output buffer to store the digest");
+    }
+
+    byte[] digestBytes = getOutput(OUT_LEN);
+    System.arraycopy(digestBytes, 0, buf, off, digestBytes.length);
+    return digestBytes.length;
+  }
+
+  @Override
+  protected void finalize() throws Throwable {
+    engineReset();
+  }
+
+  /* The following methods implement the {Hasher} interface. */
+
+  @CanIgnoreReturnValue
+  public Hasher putBytes(ByteBuffer b) {
+    buffer = b;
```

Comment:
I think we need to copy/use the argument `b` now instead of saving it for use later. I believe the contract of the interface is the caller is free to use the passed buffer after the function is returned.

Reply:
Right you are.